Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Neos\Fusion\FusionObjects\Helpers;
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Fusion\Core\Runtime;
+
+/** @Flow\Proxy(false) */
+final class LazySelfReferentialProps implements \ArrayAccess, \Stringable
+{
+    private array $valueCache = [];
+
+    private array $currentlyEvaluatingIndexes = [];
+
+    public function __construct(
+        private string $parentPath,
+        private Runtime $runtime,
+        private array $effectiveContext,
+        private string $selfReferentialId
+    ) {
+        $this->effectiveContext[$this->selfReferentialId] = $this;
+    }
+
+    public function offsetGet($path): mixed
+    {
+        if (!array_key_exists($path, $this->valueCache)) {
+            if (isset($this->currentlyEvaluatingIndexes[$path])) {
+                throw new \RuntimeException('Circular reference detected while evaluating: "' . $this->selfReferentialId . '.' . $path . '"', 1669654158);
+            }
+            $this->currentlyEvaluatingIndexes[$path] = true;
+            $this->runtime->pushContextArray($this->effectiveContext);
+            try {
+                $this->valueCache[$path] = $this->runtime->evaluate($this->parentPath . '/' . $path);
+            } finally {
+                $this->runtime->popContext();
+                unset($this->currentlyEvaluatingIndexes[$path]);
+            }
+        }
+        return $this->valueCache[$path];
+    }
```

Comment:
I'd like to suggest a slightly different mechanism to keep the responsibility of circular reference checking out of `LazySelfReferentialProps`.

First, I'd like to propose a dedicated class that encapsulates a(ny) lazy value reference:
```php
/**
 * @template T
 */
#[Flow\Proxy(false)]
final class LazyReference
{
    private bool $isLocked = false;
    private bool $hasBeenDereferenced = false;

    /**
     * @var T
     */
    private mixed $value = null;

    /**
     * @param callable():T $calculateValue
     */
    public function __construct(
        private readonly mixed $calculateValueCallback
    ) {
    }

    /**
     * @return T
     */
    public function deref(): mixed
    {
        if ($this->hasBeenDereferenced) {
            return $this->value;
        }

        if ($this->isLocked) {
            throw new \Exception('Circular reference!');
        }

        $this->isLocked = true;
        $this->value = $this->calculateValue();
        $this->hasBeenDereferenced = true;

        return $this->value;
    }

    /**
     * @return T
     */
    private function calculateValue(): mixed
    {
        $calculateValueCallback = $this->calculateValueCallback;
        return $calculateValueCallback();
    }
}
```

`LazyReference`'s `deref` method takes care of circular reference checking by locking itself upon evaluation. This mechanism can easily tested via unit test:

**Case 1: No dependencies**
```php
$lazyRef = new LazyReference(function () {
    return 'foo';
});
$lazyRef->deref(); // returns "foo"
```

**Case 2: Non-circular dependencies**
```php
$lazyRef1 = new LazyReference(function () use (&$lazyRef2) {
    return $lazyRef2->deref();
});
$lazyRef2 = new LazyReference(function () use (&$lazyRef3) {
    return $lazyRef3->deref();
});
$lazyRef3 = new LazyReference(function () {
    return 'foo';
});

$lazyRef1->deref(); // returns "foo"
```

**Case 3: Self-dependent**
```php
$lazyRef = new LazyReference(function () use (&$lazyRef) {
    return $lazyRef->deref();
});
$lazyRef->deref(); // throws "Circular reference!"
```

**Case 4: Transitive, circular dependencies**
```php
$lazyRef1 = new LazyReference(function () use (&$lazyRef2) {
    return $lazyRef2->deref();
});
$lazyRef2 = new LazyReference(function () use (&$lazyRef3) {
    return $lazyRef3->deref();
});
$lazyRef3 = new LazyReference(function () use (&$lazyRef1) {
    return $lazyRef1->deref();
});

$lazyRef1->deref(); // throws "Circular reference!"
```

*Side Note: Of course, `LazyReference` can be set up to carry an identifier, so it is able to throw a more helpful error message*

Now, with `LazyReference` in place, the `offsetGet` method of `LazySelfReferentialProps` can be implemented like this:
```php
public function offsetGet($path): mixed
{
    $fullyQualifiedPath = $this->parentPath . '/' . $path;

    $this->lazyReferences[$path] ??= new LazyReference(
        function () use ($fullyQualifiedPath) {
            $this->runtime->pushContextArray($this->effectiveContext);
            try {
                return $this->runtime->evaluate($fullyQualifiedPath);
            } finally {
                $this->runtime->popContext();
            }
        }
    );

    return $this->lazyReferences[$path]->deref();
}
```

Wdyt, @mhsdesign?

Reply:
this is most certainly the most beautifull solution thanks for you proposal ;D

Reply:
implemented ;)